drivers: adc: adm1177 : Add driver support for adm1177

RaduSabau1 opened the pull request:
Add initial header and source file for the adm1177 driver.
Signed-off-by: Radu Sabau <radu.sabau@analog.com>

## Pull Request Description

This are the drivers files for the ADM1177. It includes the next functions : 
_init - It creates the device handler, initializes the communication peripherals, requests the GPIOs and set up the device 
          according to the parameters specified in the dedicated structure.
_remove -  It frees the resources allocated by the initialization function.
_write - Writes the command byte specified through the command_byte variable to the slave address or an 
             extended command byte specified through the command byte variable to the register address specified                  
             through the reg_address variable on the slave device.
_read_status - Reads the STATUS Register of the ADM1177 and receives the status byte through the status_byte pointer.
                       It reads the correct data only if a command of status reading is written to the slave before, using the
                       _write_command_byte function.
_read - Reads both the Voltage and Current raw samples from the ADM1177's ADC and receives them through the
            conv_voltage and conv_current pointers.
            It reads the correct data only if a command of both Current and Voltage conversion is written to the slave
            before, using the _write function.
_to_microvolts - Converts the raw sample into a voltage value(mV in this case), using the formula in the data
                                    sheet and a V_FULLSCALE which is determined by the VRANGE(by default it is 0).
_to_milliamperes - Converts the raw sample into a currrent value(uA in this case) depending on the RSENSE
                              which is the value of the sense resistor in mOhm(by default is 25).

## PR Type
- [ ] Bug fix (change that fixes an issue)
- [x] New feature (change that adds new functionality)
- [ ] Breaking change (has dependencies in other repos or will cause CI to fail)

## PR Checklist
- [x] I have followed the [Coding style guidelines](http://analogdevicesinc.github.io/no-OS/drivers_guide.html#coding-style)
- [x] I have performed a self-review of the changes
- [x] I have commented my code, at least hard-to-understand parts
- [ ] I have build all projects affected by the changes in this PR
- [ ] I have tested in hardware affected projects, at the relevant boards
- [x] I have signed off all commits from this PR
- [ ] I have updated the documentation (wiki pages, ReadMe etc), if applies

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/analogdevicesinc/no-OS?pullRequest=1887) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/analogdevicesinc/no-OS?pullRequest=1887) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/analogdevicesinc/no-OS?pullRequest=1887) it.</sub>

buha commented on drivers/power/adm1177/adm1177.h:
```diff
@@ -0,0 +1,153 @@
+/* Configuration Register Definition */
+
+/* Write Command Byte Bits */
+#define V_CONT          0x1
```
Also probably it would be a good idea to prefix these with CMD_ (as you did with REG_ for registers_ to be clear they are commands so the result would look like:

`ADM1177_CMD_V_CONT`

RaduSabau1 replied:
will use no_os_bit and rename the macros.

buha commented on drivers/power/adm1177/adm1177.h:
```diff
@@ -0,0 +1,153 @@
+/******************************************************************************/
+/************************* Functions Declarations *****************************/
+/******************************************************************************/
+int8_t adm1177_init(struct adm1177_dev **device,
```
use int as return type when functions return a status code, don't use int32_t or int8_t etc.

RaduSabau1 replied:
will do.

buha commented on drivers/power/adm1177/adm1177.c:
```diff
@@ -0,0 +1,311 @@
+/***************************** Include Files **********************************/
+/******************************************************************************/
+#include <stdlib.h>
+#include "adm1177.h"    //ADM1177 Definitions
```
also remove this comment, it's clear what the include does

RaduSabau1 replied:
Removed it.

buha commented on drivers/power/adm1177/adm1177.c:
```diff
@@ -0,0 +1,311 @@
+
+	dev = (struct adm1177_dev *)no_os_malloc(sizeof(*dev));
+	if(!dev)
+		return -1;
```
#include <errno.h>

and return -ENOMEM

amiclaus replied:
I think we prefer using our `no_os_error.h` header for these drivers (it includes also errno,h). not mandatory though.

buha commented on drivers/power/adm1177/adm1177.c:
```diff
@@ -0,0 +1,311 @@
+	/* Initialize I2C peripheral */
+	status = no_os_i2c_init(&dev->i2c_desc, &init_param.i2c_init);
+	if(status)
+		return status;
```
if this i2c init fails, first deallocate the memory of dev, then return

otherwise you get mem leak

buha commented:
also, check the LinuxBuilds doxygen results, seems there are some errors in the doxygen comments

buha commented:
there are some unnecessary extra whitespaces in your file which need to be removed

amiclaus commented on drivers/power/adm1177/adm1177.c:
```diff
@@ -0,0 +1,311 @@
+ *                        Example: -1 - Initialization failed;
+ *                                  0 - Initialization succeeded.
+*******************************************************************************/
+int8_t adm1177_init(struct adm1177_dev **device,
```
use `int` type for all functions returns. check the entire code.

RaduSabau1 replied:
Will do.

amiclaus commented on drivers/power/adm1177/adm1177.c:
```diff
@@ -0,0 +1,311 @@
+		    struct adm1177_init_param init_param)
+{
+	struct adm1177_dev *dev;
+	int8_t status = -1;
```
no need for explicit initialization.

buha replied:
yeah, no need for -1 and also don't use such hardcoded return codes which don't carry more meaning than "error"

RaduSabau1 replied:
Fixing it.

amiclaus commented on drivers/power/adm1177/adm1177.c:
```diff
@@ -0,0 +1,311 @@
+ *
+ * @return 0 in case of success, negative error code otherwise.
+*******************************************************************************/
+int32_t adm1177_remove(struct adm1177_dev *device)
```
ditto: `int`

amiclaus commented on drivers/power/adm1177/adm1177.c:
```diff
@@ -0,0 +1,311 @@
+{
+	int32_t ret;
+
+	ret = no_os_i2c_read(device->i2c_desc, status_byte, 1, 1);
```
return directly. moreover, is this function `read_status` really needed? Looks like a basic i2c operation.

RaduSabau1 replied:
No need for it indeed.

amiclaus commented on drivers/power/adm1177/adm1177.c:
```diff
@@ -0,0 +1,311 @@
+ *
+ * @param dev - The device structure.
+ *
+ * @param slave_address - The slave address, by default is 0xA5.
```
you don't have this function parameter actually.

RaduSabau1 replied:
Fixing it.

amiclaus commented on drivers/power/adm1177/adm1177.c:
```diff
@@ -0,0 +1,311 @@
+	ret = no_os_i2c_read(device->i2c_desc, data, 3, 1);
+
+	if(ret)
+		return -1;
```
ditto: relevant error code.

RaduSabau1 replied:
Fixing it.

amiclaus commented on drivers/power/adm1177/adm1177.c:
```diff
@@ -0,0 +1,311 @@
+	if(ret)
+		return -1;
+
+	*conv_voltage = (data[0] << 4) | (data[2] >> 4);
```
would be nice to either add some comments explaining what you do here with respect to the data received **or** use `no_os_field_get` in conjunction with relevant masks.

amiclaus commented on drivers/power/adm1177/adm1177.c:
```diff
@@ -0,0 +1,311 @@
+	if(ret)
+		return -1;
+
+	*conv_voltage = (data[0] << 4) | (data[1] >> 4);
```
same here: either some comments, either masks usage.

amiclaus commented on drivers/power/adm1177/adm1177.c:
```diff
@@ -0,0 +1,311 @@
+	current = (((1000*105840)/4096)*(int)raw_sample)/R_SENSE;
+
+	return current;
+}
```
nit: new line at the end of the file.

RaduSabau1 replied:
Added a new line.

amiclaus commented on drivers/power/adm1177/adm1177.h:
```diff
@@ -0,0 +1,153 @@
+
+#define V_FULLSCALE_0   26.35
+#define V_FULLSCALE_1   6.65
+#define R_SENSE         25
```
whe use tabs also for macro definitions indentation.

amiclaus commented on drivers/power/adm1177/adm1177.h:
```diff
@@ -0,0 +1,153 @@
+/* Converts the raw sample of the current from the ADC in uA */
+int adm1177_current_conversion(uint16_t raw_sample);
+
+#endif /*__ADM1177_H__*/
```
nit: new line at the end of the file.

RaduSabau1 replied:
Added a new line.

amiclaus reviewed (commented):
some comments from my side. Also a generic aspect: do not use past tense when explaining what you do now. (use present tens in commit title/body)

CiprianRegus commented on drivers/power/adm1177/adm1177.c:
```diff
@@ -0,0 +1,311 @@
+	struct adm1177_dev *dev;
+	int8_t status = -1;
+
+	dev = (struct adm1177_dev *)no_os_malloc(sizeof(*dev));
```
Use `no_os_calloc()` instead

RaduSabau1 replied:
Of course.

CiprianRegus commented on drivers/power/adm1177/adm1177.c:
```diff
@@ -0,0 +1,311 @@
+			 uint16_t* conv_current)
+{
+	int32_t ret;
+
```
remove the empty line

RaduSabau1 replied:
Removed it.

CiprianRegus commented on drivers/power/adm1177/adm1177.h:
```diff
@@ -0,0 +1,153 @@
+/*************************** Types Declarations *******************************/
+/******************************************************************************/
+struct adm1177_dev {
+	/* I2C */
```
useless comment

CiprianRegus commented on drivers/power/adm1177/adm1177.c:
```diff
@@ -0,0 +1,311 @@
+{
+	int current = 0;
+
+	current = (((1000*105840)/4096)*(int)raw_sample)/R_SENSE;
```
- no need to cast to int
- add spaces between arithmetic operators.
- return directly

RaduSabau1 replied:
Didn't cast int anymore, added spaces and returned directly.

CiprianRegus commented on drivers/power/adm1177/adm1177.c:
```diff
@@ -0,0 +1,311 @@
+
+	ret = no_os_i2c_remove(device->i2c_desc);
+
+	no_os_free(device);
```
set `device` to `NULL` after the free.

RaduSabau1 commented on drivers/power/adm1177/adm1177.h:
```diff
@@ -0,0 +1,153 @@
+#include <stdio.h>
+#include "no_os_i2c.h"
+
+/******************************************************************************/
```
Will do.

RaduSabau1 commented on drivers/power/adm1177/adm1177.h:
```diff
@@ -0,0 +1,153 @@
+#define STATUS_RD       0x40
+
+/* Used to clear the Write Command Byte */
+#define CLEAR_COMMAND   0x0
```
Will remove it.

RaduSabau1 commented on drivers/power/adm1177/adm1177.c:
```diff
@@ -0,0 +1,311 @@
+	int voltage = 0;
+
+	if(v_fullscale == 0)
+		voltage = ((V_FULLSCALE_0*1000)/4096) * (int)raw_sample;
```
Casting it to uint64 multiplying by 1 million, dividing by 4096, then by 1000 and obtaining a conversion result in uV.

RaduSabau1 commented on drivers/power/adm1177/adm1177.h:
```diff
@@ -0,0 +1,153 @@
+#define OFF_STATUS      0x10
+#define OFF_ALERT       0x20
+
+#define V_FULLSCALE_0   26.35
```
Removed float, decided on the precision.

RaduSabau1 commented on drivers/power/adm1177/adm1177.h:
```diff
@@ -0,0 +1,153 @@
+
+/* SLAVE ADDRESS */
+
+#define SLAVE_ADDRESS   0x5A
```
Prefixed ever register.

RaduSabau1 commented on drivers/power/adm1177/adm1177.c:
```diff
@@ -0,0 +1,311 @@
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*******************************************************************************/
+
+/******************************************************************************/
```
sure thing.

RaduSabau1 commented on drivers/power/adm1177/adm1177.h:
```diff
@@ -0,0 +1,153 @@
+#define EN_OFF_ALERT    0x8
+#define CLEAR           0x10
+
+/* CONTROL Register Operations */
```
on it.

buha commented on drivers/power/adm1177/adm1177.h:
```diff
@@ -0,0 +1,133 @@
+int adm1177_remove(struct adm1177_dev *device);
+
+/* Write the command byte to the slave */
+int adm1177_write_command_byte(struct adm1177_dev *device,
```
To simplify the API, you can combine the adm1177_write_command_byte with adm1177_write_extended_command_byte.

In the merged function (could be simply called `adm1177_write(dev, cmd, data)`) check for the reg_address's MSB and write just the cmd or the cmd+data (1 or 2 bytes write).

RaduSabau1 replied:
I will merge them, but use a slightly different way. Will call teh function adm1177_write(device, cmd,reg_address) and will check if the reg_address's MSB is 1, then perform a write extended command byte operation, otherwise a simple write operation.

buha commented on drivers/power/adm1177/adm1177.h:
```diff
@@ -0,0 +1,133 @@
+			uint8_t *status_byte);
+
+/* Gets conversion results for both Voltage and Current at the same time */
+int adm1177_readback(struct adm1177_dev *device,
```
readback, voltage_readback, and current_readback functions could all be just 1 single API function - readback.

Use `dev` to store information about what channels are enabled, and internally use that information to know when to read just voltage, just current or both.

RaduSabau1 replied:
Will do, and also modify the channel enabled when a command is written if the command corresponds to a voltage, current or both reading.

buha commented on drivers/power/adm1177/adm1177.h:
```diff
@@ -0,0 +1,133 @@
+			     uint16_t* conv_current);
+
+/* Converts the raw sample of the voltage from the ADC in mV */
+int adm1177_voltage_conversion(uint16_t raw_sample,
```
i'm ok with `voltage_conversion` name, but could be even more explicit - `to_millivolts`

RaduSabau1 replied:
Will change.

buha commented on drivers/power/adm1177/adm1177.c:
```diff
@@ -0,0 +1,294 @@
+#include "no_os_alloc.h"
+#include "no_os_delay.h"
+#include "no_os_error.h"
+/**************************************************************************//**
```
maybe my comments about blank lines have previously misled you and i see you removed too many blank lines now :D

have a look at how this file looks and try to match the style, it's easier to put it this way:
https://github.com/analogdevicesinc/no-OS/blob/master/drivers/cdc/ad7746/ad7746.c

buha replied:
the reason for this blank line and white space nitpicking is consistency, we want to code to also look homogeneous

RaduSabau1 replied:
Will try to match the style.

buha commented on drivers/power/adm1177/adm1177.c:
```diff
@@ -0,0 +1,294 @@
+		return (((uint64_t)ADM1177_V_FULLSCALE_0 * 1000000) /
+			4096) / 1000 * raw_sample;
+	default:
+		return -1;
```
avoid -1, pick a meaningful error code from errno.h like -EINVAL in this case

RaduSabau1 replied:
Will do.

RaduSabau1 commented on drivers/power/adm1177/adm1177.h:
```diff
@@ -0,0 +1,133 @@
+#define ADM1177_CMD_STATUS_RD   	NO_OS_BIT(6)
+
+/* ADM11777 Write Extended Command Byte Register Addresses */
+#define ADM1177_REG_ALERT_EN    	NO_OS_BIT(7)
```
Indeed, will change.

amiclaus commented on drivers/power/adm1177/adm1177.h:
```diff
@@ -0,0 +1,124 @@
+int adm1177_init(struct adm1177_dev **device,
+		 struct adm1177_init_param *init_param);
+
+int adm1177_remove(struct adm1177_dev *device);
```
either add header comment for all function prototypes or none of them.

RaduSabau1 replied:
Will add header comment for all function prototypes.

amiclaus commented on drivers/power/adm1177/adm1177.c:
```diff
@@ -0,0 +1,254 @@
+		return (((uint64_t)ADM1177_V_FULLSCALE_0 * 1000000) /
+			4096) / 1000 * raw_sample;
+	case 1:
+		return (((uint64_t)ADM1177_V_FULLSCALE_0 * 1000000) /
```
should be `ADM1177_V_FULLSCALE_1` here or am I missing something?

RaduSabau1 replied:
Modifying it.

amiclaus commented on drivers/power/adm1177/adm1177.c:
```diff
@@ -0,0 +1,254 @@
+			  int v_fullscale)
+{
+	switch(v_fullscale) {
+	case 0:
```
an enum might fit here for the available modes.

RaduSabau1 replied:
Made an enum.

amiclaus commented on drivers/power/adm1177/adm1177.c:
```diff
@@ -0,0 +1,254 @@
+	switch(device->chan_en) {
+	case ADM1177_CURRENT_EN:
+		ret = no_os_i2c_read(device->i2c_desc, data, 2, 1);
+
```
remove extra line

amiclaus commented on drivers/power/adm1177/adm1177.c:
```diff
@@ -0,0 +1,254 @@
+		return 0;
+	case ADM1177_VOLTAGE_EN:
+		ret = no_os_i2c_read(device->i2c_desc, data, 2, 1);
+
```
ditto

amiclaus commented on drivers/power/adm1177/adm1177.c:
```diff
@@ -0,0 +1,254 @@
+		return 0;
+	case ADM1177_VOLTAGE_AND_CURRENT_EN:
+		ret = no_os_i2c_read(device->i2c_desc, data, 3, 1);
+
```
ditto

amiclaus commented on drivers/power/adm1177/adm1177.c:
```diff
@@ -0,0 +1,254 @@
+
+	*device = dev;
+
+	return status;
```
you are returning a int8_t type variable for a function that hast `int` return type.

RaduSabau1 replied:
Will change.

amiclaus commented on drivers/power/adm1177/adm1177.c:
```diff
@@ -0,0 +1,254 @@
+	switch(v_fullscale) {
+	case 0:
+		return (((uint64_t)ADM1177_V_FULLSCALE_0 * 1000000) /
+			4096) / 1000 * raw_sample;
```
what is 4096? looks like a resolution. Map to a suggestive macro. Same applies for 1000, 1000000. I think you might want to use `no_os_units.h`, more exactly:
https://github.com/analogdevicesinc/no-OS/blob/4c555a808daf2ef2e7e0c2034259fb5780b6f1b9/include/no_os_units.h#L65

RaduSabau1 replied:
Will use them, thanks.

amiclaus commented on drivers/power/adm1177/adm1177.c:
```diff
@@ -0,0 +1,254 @@
+ * @return - Conversion result in uV.
+ * 	     -1 if v_fullscale is set wrong.
+******************************************************************************/
+int adm1177_to_microvolts(uint16_t raw_sample,
```
use pointer to store the output value and pass it as parameter. you are currently returning a mix of error codes and uV values (uint64_t type) on a `int` type function.

RaduSabau1 replied:
Will use pointer to store output value.

amiclaus commented on drivers/power/adm1177/adm1177.c:
```diff
@@ -0,0 +1,254 @@
+******************************************************************************/
+int adm1177_to_microampers(uint16_t raw_sample)
+{
+	return (((1000 * 105840) / 4096) * raw_sample) / ADM1177_R_SENSE;
```
can you have here negative values returned? If not, adjust the function return type accordingly or use pointers to store the output value.

RaduSabau1 replied:
Adjusting the function return type.

amiclaus commented on drivers/power/adm1177/adm1177.c:
```diff
@@ -0,0 +1,254 @@
+
+		return no_os_i2c_write(device->i2c_desc, data, 2, 1);
+	} else {
+		if(cmd == ADM1177_CMD_V_CONT | cmd == ADM1177_CMD_V_ONCE)
```
please use `switch...case`. Improves code readability.

RaduSabau1 replied:
Will do.

RaduSabau1 commented on drivers/power/adm1177/adm1177.h:
```diff
@@ -0,0 +1,124 @@
+
+struct adm1177_dev {
+	struct no_os_i2c_desc   *i2c_desc;
+
```
Will remove.

RaduSabau1 commented on drivers/power/adm1177/adm1177.c:
```diff
@@ -0,0 +1,254 @@
+		  uint8_t cmd,
+		  uint8_t reg_address)
+{
+	if(reg_address & NO_OS_BIT(7)) {
```
Will map indeed.

RaduSabau1 commented on drivers/power/adm1177/adm1177.h:
```diff
@@ -0,0 +1,124 @@
+/* Status Byte Operations */
+#define ADM1177_ADC_OC_MASK    		NO_OS_BIT(0)
+#define ADM1177_ADC_ALERT_MASK 		NO_OS_BIT(1)
+#define ADM1177_HS_OC_MASK     		NO_OS_BIT(2)
```
Solving it.

RaduSabau1 commented on drivers/power/adm1177/adm1177.h:
```diff
@@ -0,0 +1,124 @@
+#define ADM1177_OFF_STATUS_MASK		NO_OS_BIT(4)
+#define ADM1177_OFF_ALERT_MASK 		NO_OS_BIT(5)
+
+#define ADM1177_V_FULLSCALE_0   26350
```
Solving it.

amiclaus reviewed (commented):
some more comments on my side. Also, when I referred to the commit body/title addressing mode, please check: https://git.kernel.org/pub/scm/git/git.git/tree/Documentation/SubmittingPatches?h=v2.36.1#n181

amiclaus reviewed (commented):
noticed that you changed the location of the driver under `power` . Looking at the linux driver from the master branch, the device might fit either under `hwmon` or `adc`:
https://github.com/analogdevicesinc/linux/blob/master/drivers/iio/adc/adm1177.c
https://github.com/analogdevicesinc/linux/blob/master/drivers/hwmon/adm1177.c

RaduSabau1 commented on doc/doxygen/CMakeLists.txt:
```diff
@@ -18,7 +18,7 @@ set(DOCUMENTED_FILES
     ${CMAKE_CURRENT_SOURCE_DIR}/mainpage.dox \\
     ${SOURCES_DIR}/include \\
-    ${SOURCES_DIR}/iio \\
+    ${SOURCES_DIR}/iio	\\
```
Will remove.

RaduSabau1 commented on drivers/power/adm1177/adm1177.c:
```diff
@@ -0,0 +1,277 @@
+			device->chan_en = ADM1177_VOLTAGE_EN;
+			break;
+		case ADM1177_CMD_V_ONCE:
+			device->chan_en = ADM1177_VOLTAGE_EN;
```
Replacing.

RaduSabau1 commented on drivers/power/adm1177/adm1177.c:
```diff
@@ -0,0 +1,277 @@
+
+		return 0;
+	case ADM1177_VOLTAGE_EN:
+		ret = no_os_i2c_read(device->i2c_desc, data, 2, 1);
```
Can't do a fall through over here, because the operation is the same but a different variable is changed.

amiclaus commented on drivers/adc/adm1177/adm1177.c:
```diff
@@ -0,0 +1,274 @@
+
+	no_os_free(device);
+
+	device = NULL;
```
please remove. This is redundant in the way we pass pointers in the `_remove` function.

RaduSabau1 replied:
Will remove.

amiclaus commented on drivers/adc/adm1177/adm1177.c:
```diff
@@ -0,0 +1,274 @@
+
+	uint8_t data[3];
+
+	switch(device->chan_en) {
```
i'd approach this readback function similar to https://github.com/analogdevicesinc/linux/blob/9e1304f0416a5bd3c79c48bc73d24bf0978a47d7/drivers/iio/adc/adm1177.c#L92-L103

I'd readback only one value based on the user request (either voltage or current).

The current approach might introduce erroneous read-backs in the upper layers. For case `ADM1177_CURRENT_EN ` the content of `conv_voltage` might be a older reading/ garbage value (but without any warning to the user) same for case `ADM1177_VOLTAGE_EN` and `conv_current`.

RaduSabau1 replied:
I understand what you're saying, but I'm not really sure how I can approach it similar to the linux driver, in the way that if i write a command to the ADM saying I only need to read current then the current data comes in a different form, same for the voltage. So in my opinion I need that third case. 
What I can do is for the Voltage only readback and Current only readback, to make the other one 0 and mention it in the documentation, this way no erroneous read-backs will appear in the upper layers.

amiclaus replied:
Or you can add some wrappers. Rename this function `adm1177_readback_raw` and add some wrappers where you interpret the data `adm1177_read_voltage_raw` `adm1177_read_current_raw` `adm1177_read_voltage_and_current_raw`. Just a thought.

amiclaus commented on drivers/adc/adm1177/iio_adm1177.c:
```diff
@@ -0,0 +1,170 @@
+
+	*iio_dev = desc;
+
+	no_os_mdelay(15);
```
why delay? why 15ms?

RaduSabau1 replied:
It is mentioned in the data sheet that the ADM1177 needs a specific time after intialization before performing any other operation.

RaduSabau1 replied:
Ok then.

buha replied:
remove this delay as it's the initial timing cycle delay from poweron and the driver has no control whatsoever of the poweron so it can't know how much time has passed from poweron to driver init...

amiclaus commented on drivers/adc/adm1177/iio_adm1177.c:
```diff
@@ -0,0 +1,170 @@
+	ret = adm1177_write(desc->adm1177_dev, ADM1177_CMD_V_CONT |
+			    ADM1177_CMD_I_CONT, 0);
+	if(ret)
+		return ret;
```
`goto error_desc;`

RaduSabau1 replied:
On it.

amiclaus commented on drivers/adc/adm1177/iio_adm1177.h:
```diff
@@ -0,0 +1,59 @@
+};
+
+struct adm1177_iio_init_param {
+	struct adm1177_init_param *adm1177_initial;
```
nit: i'd rename to `adm1177_iio_init`

amiclaus commented on drivers/adc/adm1177/iio_adm1177.h:
```diff
@@ -0,0 +1,59 @@
+#include "iio.h"
+
+struct adm1177_iio_dev {
+	struct adm1177_dev *adm1177_dev;
```
nit: i'd rename to `adm1177_iio_dev`

amiclaus replied:
you can neglect this comment on my side.

RaduSabau1 replied:
In my opinion that would make the code a bit confusing, especially in the _iio_init function.

amiclaus commented on drivers/adc/adm1177/iio_adm1177.c:
```diff
@@ -0,0 +1,170 @@
+
+	switch(channel->type) {
+	case IIO_VOLTAGE:
+		ret = adm1177_readback(desc, &conv_voltage, &conv_current);
```
this code si duplicated for both `IIO_VOLTAGE` and `IIO_CURRENT`

RaduSabau1 replied:
Not really. The read value is transmitted through a different variable(in IIO_VOLTAGE case through conv_voltage, and in IIO_CURRENT through conv_current).

amiclaus reviewed (commented):
some more comments on my side. Also please provide changelogs after each force-push. (example: https://github.com/analogdevicesinc/no-OS/pull/1888#pullrequestreview-1613437966)

buha commented on drivers/adc/adm1177/adm1177.c:
```diff
@@ -0,0 +1,282 @@
+		   shifted by 4, and when the OR operation between the
+		   two bytes happen we obtain the a 12-bit value of the
+		   current. */
+		*conv_current = (data[0] << 4) | (data[1] >> 4);
```
if the user calls the function with adm1177_readback(dev, NULL, ptr) your program will crash because you are dereferencing NULL pointer

```
if (conv_current == NULL)
    return -EINVAL;

*conv_current = ...;
    
```

RaduSabau1 replied:
Will change, and will add a specific function for read status.

buha replied:
move this test before the i2c_read as well

buha commented on drivers/adc/adm1177/adm1177.c:
```diff
@@ -0,0 +1,316 @@
+		   byte by 4 we are shifting the 8 MSBs of the current
+		   by 4, and adding the third byte masked by 0xF
+		   we add the remaining 4 LSBs of the current.*/
+		*conv_voltage = (data[0] << 4) | (data[2] >> 4);
```
missing non NULL check for both conv_current and conv_voltage

RaduSabau1 replied:
Will add.

buha commented on drivers/adc/adm1177/adm1177.c:
```diff
@@ -0,0 +1,316 @@
+ *
+ * @return 0 in case of success, negative error code otherwise.
+******************************************************************************/
+int adm1177_readback(struct adm1177_dev *device,
```
i think `adm1177_read` is a better fit to complement existing API functions `adm1177_write` and `adm1177_read_status`

RaduSabau1 replied:
Will change.

buha commented on drivers/adc/adm1177/adm1177.c:
```diff
@@ -0,0 +1,316 @@
+{
+	switch(v_fullscale) {
+	case ADM1177_VRANGE_LOW:
+		*voltage_uv = (((uint64_t)ADM1177_V_FULLSCALE_0 *
```
it would probably make sense that this function has the signature:
```
int adm1177_to_microvolts(dev, raw_in, volts_out)
```

And within the function you use the dev->vrange to perform the computation.

RaduSabau1 replied:
Will do.

RaduSabau1 commented on drivers/adc/adm1177/adm1177.c:
```diff
@@ -0,0 +1,316 @@
+		if(ret)
+			return ret;
+
+		if(conv_voltage == NULL)
```
Of course.

buha commented on drivers/adc/adm1177/adm1177.c:
```diff
@@ -0,0 +1,315 @@
+			device->last_command = ADM1177_VOLTAGE_EN;
+			return no_os_i2c_write(device->i2c_desc, &cmd, 1, 1);
+		}
+		if(((cmd == ADM1177_CMD_I_CONT | ADM1177_CMD_V_CONT)) ||
```
== operator has higher precedence than | and because of this you have a bug here

put all conditions in parentheses

RaduSabau1 replied:
will do

buha commented on drivers/adc/adm1177/adm1177.c:
```diff
@@ -0,0 +1,315 @@
+	if(device->last_command == ADM1177_STATUS_EN) {
+		if(status_byte == NULL)
+			return -EINVAL;
+		int ret;
```
move variable declarations at the beginning of function, I'm pretty sure this is in the style guide

RaduSabau1 replied:
ok.

buha commented:
- [x] implement the iio SCALE
- [x] implement the iio buffer (in my opinion optional, but would be nice to have)

amiclaus commented on drivers/adc/adm1177/adm1177.c:
```diff
@@ -0,0 +1,315 @@
+			return ret;
+
+		return 0;
+	} else {
```
this else statement can be dropped. Doesn't add much value.

amiclaus commented on drivers/adc/adm1177/adm1177.c:
```diff
@@ -0,0 +1,315 @@
+		uint8_t data[2] = {reg_address, cmd};
+
+		return no_os_i2c_write(device->i2c_desc, data, 2, 1);
+	} else {
```
ditto: this else statement can be dropped. Doesn't add much value.

amiclaus commented on drivers/adc/adm1177/iio_adm1177.c:
```diff
@@ -0,0 +1,168 @@
+
+	desc = (struct adm1177_iio_dev *)no_os_calloc(1, sizeof(*desc));
+	if(!desc)
+		return desc;
```
do we want to return success corresponding code here?

amiclaus commented on drivers/adc/adm1177/iio_adm1177.c:
```diff
@@ -0,0 +1,168 @@
+	if(ret)
+		goto error_desc;
+
+	ret = adm1177_write(desc->adm1177_dev, ADM1177_CMD_V_CONT |
```
instead of making only the iio driver ready for curren/voltage reads why not also the main driver? (by moving thins operation in `adm1177_init`

amiclaus commented on drivers/adc/adm1177/adm1177.c:
```diff
@@ -0,0 +1,320 @@
+		return no_os_i2c_write(device->i2c_desc, &cmd, 1, 1);
+	}
+	if(cmd & ADM1177_CMD_VRANGE) {
+		device->vrange = ADM1177_VRANGE_HIGH;
```
you are returning an error code after this assignment. do we want that?

amiclaus replied:
Where do we set the voltage range in hardware actually?
![image](https://github.com/analogdevicesinc/no-OS/assets/26061529/bb054ea1-2323-4483-b832-257b76169f68)

RaduSabau1 replied:
I see what you mean now. And can see why this is a bug, fixing it now.

amiclaus commented on drivers/adc/adm1177/adm1177.c:
```diff
@@ -0,0 +1,320 @@
+	if(cmd & ADM1177_CMD_VRANGE) {
+		device->vrange = ADM1177_VRANGE_HIGH;
+	} else {
+		device->vrange = ADM1177_VRANGE_LOW;
```
ditto: you are returning an error code after this assignment. do we want that?

amiclaus commented on drivers/adc/adm1177/adm1177.c:
```diff
@@ -0,0 +1,320 @@
+ * @param init_param - The structure that contains the device initial
+ * 		       parameters.
+ *
+ * @return status       - Initialization status.
```
odd indentation.

buha commented on drivers/adc/adm1177/iio_adm1177.c:
```diff
@@ -0,0 +1,251 @@
+		if(adm1177->vrange == ADM1177_VRANGE_LOW)
+			vals[0] = ADM1177_V_FULLSCALE_0;
+		else {
+			vals[1] = ADM1177_V_FULLSCALE_1;
```
should be vals[0], this is a bug

RaduSabau1 replied:
On it

buha commented on drivers/adc/adm1177/iio_adm1177.c:
```diff
@@ -0,0 +1,251 @@
+		return iio_format_value(buf, len, IIO_VAL_FRACTIONAL_LOG2, 2,
+					vals);
+	case IIO_CURRENT:
+		vals[0] = 105840;
```
use ADM1177_I_FULLSCALE

RaduSabau1 replied:
Of course.

buha commented on drivers/adc/adm1177/iio_adm1177.c:
```diff
@@ -0,0 +1,251 @@
+	adm1177 = iio_adm1177->adm1177_dev;
+
+	for(uint32_t i = 0; i < samples*2; i++) {
+		adm1177_read(adm1177, &conv_voltage, &conv_current);
```
check return status

RaduSabau1 replied:
Ok.

buha commented on drivers/adc/adm1177/iio_adm1177.c:
```diff
@@ -0,0 +1,251 @@
+
+	adm1177 = iio_adm1177->adm1177_dev;
+
+	for(uint32_t i = 0; i < samples*2; i++) {
```
In this loop you need to check if the channel is enabled and populate samples only if it's enabled:
https://github.com/analogdevicesinc/no-OS/blob/master/drivers/accel/adxl355/iio_adxl355.c#L933

The driver underneath and the chip, they will work with the assumption of V_CONT + I_CONT all the time, but IIO wants to be able to enable/disable channels so you'll have to do that.

buha commented on drivers/adc/adm1177/iio_adm1177.c:
```diff
@@ -0,0 +1,251 @@
+	END_ATTRIBUTES_ARRAY
+};
+
+static struct iio_attribute adm1177_iio_current_attrs[] = {
```
i don't think you need a separate variable since the voltage_attrs is identical

RaduSabau1 replied:
That is right.

RaduSabau1 reviewed (commented):
Requested changes made.

CiprianRegus commented on drivers/adc/adm1177/iio_adm1177.c:
```diff
@@ -0,0 +1,270 @@
+
+	adm1177 = iio_adm1177->adm1177_dev;
+
+	for(uint32_t i = 0; i < samples*iio_adm1177->no_of_active_channels;) {
```
What's the sampling rate of this device? Are you sure you can read individual samples so fast?

CiprianRegus commented on drivers/adc/adm1177/iio_adm1177.c:
```diff
@@ -0,0 +1,270 @@
+		ret = adm1177_read(adm1177, &conv_voltage, &conv_current);
+		if(ret)
+			return ret;
+		raw_voltage = (uint32_t)conv_voltage;
```
Drop the cast

CiprianRegus commented on drivers/adc/adm1177/iio_adm1177.c:
```diff
@@ -0,0 +1,270 @@
+};
+
+enum adm1177_chan {
+	VOLTAGE,
```
Add the `ADM1177_` prefix, you might get name conflicts in the future.

RaduSabau1 replied:
Will add.

CiprianRegus commented on drivers/adc/adm1177/iio_adm1177.c:
```diff
@@ -0,0 +1,270 @@
+		raw_current = (uint32_t)conv_current;
+
+		if(iio_adm1177->active_channels & NO_OS_BIT(0)) {
+			buff[i] = no_os_sign_extend32(raw_voltage, 11);
```
You set the `sign` field in the `scan_type` to `'u'`. Why do you need the sign extension for?

RaduSabau1 replied:
That's right.

CiprianRegus commented on drivers/adc/adm1177/iio_adm1177.c:
```diff
@@ -0,0 +1,270 @@
+#include "no_os_error.h"
+#include "no_os_util.h"
+#include "no_os_alloc.h"
+#include "no_os_delay.h"
```
You're not using this header

RaduSabau1 replied:
Removing it.

RaduSabau1 commented on drivers/adc/adm1177/iio_adm1177.c:
```diff
@@ -0,0 +1,270 @@
+			vals[0] = ADM1177_V_FULLSCALE_0;
+		else {
+			vals[0] = ADM1177_V_FULLSCALE_1;
+		}
```
Of course.

RaduSabau1 commented on drivers/adc/adm1177/iio_adm1177.c:
```diff
@@ -0,0 +1,270 @@
+
+	iio_adm1177->active_channels = mask;
+
+	if(mask & NO_OS_BIT(0))
+		counter++;
+	if(mask & NO_OS_BIT(1))
+		counter++;
```
Will try it.

buha commented on drivers/adc/adm1177/iio_adm1177.c:
```diff
@@ -0,0 +1,263 @@
+			return ret;
+
+		value = conv_voltage;
+
```
i know it's a detail but this blank line bothers me here

buha commented on drivers/adc/adm1177/iio_adm1177.c:
```diff
@@ -0,0 +1,263 @@
+			return ret;
+
+		value = conv_current;
+
```
same here

buha commented on drivers/adc/adm1177/iio_adm1177.c:
```diff
@@ -0,0 +1,263 @@
+
+	adm1177 = iio_adm1177->adm1177_dev;
+
+	for(uint32_t i = 0; i < samples*iio_adm1177->no_of_active_channels;) {
```
i am not sure but it seems that the condition is wrong, given that you read 2 channels at the time (with the device driver)

if you have 2 channels enabled, the loop will execute twice, whereas it should've executed only once because your adm1177_read gives you both voltage and current

i think you can drop the for loop

RaduSabau1 replied:
Will try it with a while loop.

buha commented on drivers/adc/adm1177/adm1177.c:
```diff
@@ -0,0 +1,350 @@
+
+	*device = dev;
+
+	status = adm1177_write(&dev->i2c_desc, ADM1177_CMD_V_CONT |
```
I would move this out of here, let the user decide when they wanna start converting.
- in app code
- in adm1177_iio_init (you will need to move this line there)

RaduSabau1 commented:
V2

-  Removed unnecessary delay in the initalization of the iio_device.
-  Added a specific function _set_vrange in order to set the vrange without affecting the last command.
-  iio driver is buffer capable, has iio scale implemented, and _update channel function is also implemented.

RaduSabau1 commented:
V3

- Removed unnecessary blank spaces.
- Removed unused headers.

buha commented:
looks good now, waiting for a 2nd approval

amiclaus commented on drivers/adc/adm1177/adm1177.c:
```diff
@@ -0,0 +1,341 @@
+ *
+ * @return 0 in case of success, negative error code otherwise.
+******************************************************************************/
+int adm1177_read(struct adm1177_dev *device,
```
`adm1177_read` means that it is a generic read function usable for any kind of register reads. which is not the case here. a more specific header should be used (same as we have for `adm1177_read_status`).

RaduSabau1 replied:
other than readback, which was the previous name for this read function doesn't come to mind after reading the data sheet.

amiclaus commented on drivers/adc/adm1177/iio_adm1177.c:
```diff
@@ -0,0 +1,268 @@
+					vals);
+	case IIO_CURRENT:
+		vals[0] = ADM1177_I_FULLSCALE;
+		vals[1] = 4096 * ADM1177_R_SENSE;
```
we already have a macro defined for 4096. we should use it.

amiclaus commented on drivers/adc/adm1177/iio_adm1177.c:
```diff
@@ -0,0 +1,268 @@
+			vals[0] = ADM1177_V_FULLSCALE_0;
+		else
+			vals[0] = ADM1177_V_FULLSCALE_1;
+		vals[1] = 12;
```
shouldn't we have something like: `vals[1] = adm1177_iio_scan_type.realbits` ?

amiclaus commented on drivers/adc/adm1177/iio_adm1177.c:
```diff
@@ -0,0 +1,268 @@
+
+	adm1177 = iio_adm1177->adm1177_dev;
+
+	uint32_t i = 0;
```
all variable declaration should be placed a the start of the function.

amiclaus commented on drivers/adc/adm1177/iio_adm1177.c:
```diff
@@ -0,0 +1,268 @@
+
+static int adm1177_iio_read_samples(void *device, int *buff, uint32_t samples)
+{
+	uint16_t conv_voltage;
```
`uint16_t conv_voltage, conv_current;`
`uint32_t raw_voltage, raw_current;`

amiclaus commented on drivers/adc/adm1177/iio_adm1177.c:
```diff
@@ -0,0 +1,268 @@
+	adm1177 = iio_adm1177->adm1177_dev;
+
+	uint32_t i = 0;
+	while(i < samples*iio_adm1177->no_of_active_channels) {
```
spaces before and after `*` operator.

amiclaus commented on drivers/adc/adm1177/iio_adm1177.c:
```diff
@@ -0,0 +1,268 @@
+	uint32_t i = 0;
+	while(i < samples*iio_adm1177->no_of_active_channels) {
+		ret = adm1177_read(adm1177, &conv_voltage, &conv_current);
+		if(ret)
```
`if (ret)` space between if and condition. check entire code for these occurrences.

amiclaus replied:
same for `while (`, `for (`, etc.

amiclaus replied:
These comments were not fully addressed.

RaduSabau1 replied:
Changing them.

RaduSabau1 commented:
V4

- Changed name of the _read function to _read_conv.
- Added space between if and condition.
- Moved variable declaration at the start of the function.

amiclaus commented on drivers/adc/adm1177/iio_adm1177.c:
```diff
@@ -0,0 +1,266 @@
+
+	switch (channel->type) {
+	case IIO_VOLTAGE:
+		ret = adm1177_read(desc, &conv_voltage, &conv_current);
```
now this function doesn't exist anymore. please replace all `adm1177_read` with the new function header.

RaduSabau1 commented:
V6

- Added space between if, switch, while, for and the conditions.

amiclaus commented on drivers/adc/adm1177/adm1177.c:
```diff
@@ -0,0 +1,341 @@
+		if (conv_current == NULL)
+			return -EINVAL;
+
+		no_os_udelay(150);
```
i suppose this is the ADC Conversion time from the datasheet specs. Why duplicating it 3 times instead of moving it before the switch case? Also, a comment would be nice for this "magic" number. See: https://github.com/analogdevicesinc/no-OS/blob/c9e00645575f1a02d46dfbb72cbf283324f6d265/drivers/adc-dac/ad74416h/ad74416h.c#L872-L873

RaduSabau1 replied:
Will handle it.

amiclaus reviewed (commented):
looks good, one more thing on my side.

RaduSabau1 commented:
V7

- Moved the delay needed for the device conversion time before the switch.
- Commented why the 150us value is used.

RaduSabau1 commented:
V8

- Rebased with master.

amiclaus reviewed (approved):
lgtm
